File: pyps/src/main/scala/com/tencent/angel/common/Meta.java
package com.tencent.angel.common;

import com.tencent.angel.matrix.MatrixContext;
import com.tencent.angel.ml.servingmath2.utils.RowType;
import com.tencent.angel.model.MatrixLoadContext;
import com.tencent.angel.model.MatrixSaveContext;
import com.tencent.angel.ps.tensor.TensorMeta;


import java.util.Arrays;


public abstract class Meta {
    protected String name;
    protected String dtype;
    protected int dim;
    protected long[] shape;
    protected long validIndexNum;
    protected RowType rowType;
    protected MatrixContext matrixContext;

    public Meta(String name, String dtype, int dim, long[] shape, long validIndexNum) {
        this.name = name;
        this.dtype = dtype;
        this.dim = dim;
        this.shape = shape;
        this.validIndexNum = validIndexNum;
        this.rowType = Utils.getRowType(dtype);
    }

    public String getName() {
        return name;
    }

    public String getDtype() {
        return dtype;
    }

    public int getDim() {
        return dim;
    }

    public long[] getShape() {
        return shape;
    }

    public long getValidIndexNum() {
        return validIndexNum;
    }

    public RowType getRowType() {
        return rowType;
    }

    public abstract MatrixContext getMatrixContext();

    public abstract MatrixSaveContext getMatrixSaveContext(String path, String formatClassName);

    public abstract MatrixLoadContext getMatrixLoadContext(String path);

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        } else if (obj instanceof TensorMeta) {
            TensorMeta other = (TensorMeta) obj;

            if (!this.name.equalsIgnoreCase(other.name)) {
                return false;
            }

            if (this.dim != other.dim) {
                return false;
            }

            if (!Arrays.equals(this.shape, other.shape)) {
                return false;
            }

            if (!this.dtype.equalsIgnoreCase(other.dtype)) {
                return false;
            }

            return this.validIndexNum == other.validIndexNum;
        } else {
            return false;
        }
    }

}

File: pyps/src/main/scala/com/tencent/client/common/Meta.scala
package com.tencent.client.common

import java.util

import com.tencent.angel.matrix.MatrixContext
import com.tencent.angel.ml.servingmath2.utils.RowType
import com.tencent.angel.model.MatrixLoadContext
import com.tencent.angel.model.MatrixSaveContext
import com.tencent.client.ps.tensor.TensorMeta


abstract class Meta(val name: String, val dtype: String, val dim: Int, val shape:Array[Long], val validIndexNum: Long) {
    val rowType: RowType = Utils.getRowType(dtype)
    protected val matrixContext: MatrixContext

    def getMatrixContext: MatrixContext

    def getMatrixSaveContext(path: String, formatClassName: String): MatrixSaveContext

    def getMatrixLoadContext(path: String): MatrixLoadContext

    override def equals(obj: Any): Boolean = {
        if (this == obj) {
            true
        } else if (obj.isInstanceOf[TensorMeta]) {
            val other = obj.asInstanceOf[TensorMeta]

            if (!this.name.equalsIgnoreCase(other.name)) {
                return false
            }

            if (this.dim != other.dim) {
                return false
            }

            if (!util.Arrays.equals(this.shape, other.shape)) {
                return false
            }

            if (!this.dtype.equalsIgnoreCase(other.dtype)) {
                return false
            }

            this.validIndexNum == other.validIndexNum
        } else {
            false
        }
    }

}
